### ISSUE TITLE
Add lint for impl-only-use RFC

### ISSUE THREAD

**Issue by mathstuf** [rust-lang/rust-clippy#2476]
Detecting the case where the [impl-only-use RFC](https://github.com/rust-lang/rfcs/pull/2166) allows for `use Trait as _` would be nice to have to find existing uses of `use Trait as UnusedName`.

**Comment by nk9** [comment]
I'm afraid I'm having trouble understanding what this issue is proposing. Can you please clarify?

@rustbot label S-waiting-on-author A-lint

**Comment by samueltardieu** [comment]
A trait can be made visible using any of the following construct:

```rust
use Trait;
use Trait as _;
use Trait as SomeOtherName;
```

If I understand this issue correctly, it proposes to lint against the third form if `SomeOtherName` is never used explictly.

@rustbot label -S-waiting-on-author

**Comment by y21** [comment]
This sounds a bit like https://rust-lang.github.io/rust-clippy/master/index.html#unused_trait_names ? (Both `use Trait` and `use Trait as UnusedName` would trigger the lint if the imported name is unused and it's only imported for its methods)

**Comment by samueltardieu** [comment]
~~It won't lint against the horrendous:~~ (my bad, it does lint)

```rust
use std::io::Read as foo;

fn main() -> Result<(), std::io::Error> {
    let mut file = std::fs::File::open("/tmp/foo")?;
    let s = &mut vec![];
    println!("{}", file.read_to_end(s)?);
    Ok(())
}
```

**Comment by mathstuf** [comment]
Yes, @samueltardieu has it exactly right above. I'm interested in finding instances where I gave a name for a trait on import (IIRC, due to async/std `Write` traits conflicting) and can use the `_` for one of them. Since it seems like it is already fixed, I'm guessing I'm using the renames *somewhere* in the case(s) I was worried about. Or they're in projects without active CI schedules :) .
